bundles/org.eclipse.swt/Eclipse SWT Browser/win32/org/eclipse/swt/browser/Edge.java

```diff
@@ -700,6 +704,60 @@ void browserResize(Event event) {
 	controller.put_IsVisible(true);
 }
 
+private void handleBrowserHover() {
+	browser.getDisplay().timerExec(100, () -> {
+		if (browser.isDisposed()) {
+			return;
+		}
+		if (!browser.isVisible() || !hasDisplayFocus()) {
+			handleBrowserHover();
+			return;
+		}
+
+		final Point cursorLocation = browser.getDisplay().getCursorLocation();
+		Point cursorLocationInControlCoordinate = browser.toControl(cursorLocation);
+		boolean isCursorInsideBrowser = browser.getBounds().contains(cursorLocationInControlCoordinate);
+		boolean isCursorLocationChanged = !cursorLocation.equals(previousCursorLocation);
+		previousCursorLocation = cursorLocation;
+
+		switch (cursorPosition) {
+		case INSIDE:
+			if (isCursorInsideBrowser && isCursorLocationChanged) {
+				sendMouseEvent(cursorLocationInControlCoordinate, SWT.MouseMove);
+			} else if (!isCursorInsideBrowser) {
+				cursorPosition = CursorPosition.OUTSIDE;
+				sendMouseEvent(cursorLocationInControlCoordinate, SWT.MouseExit);
+			}
+			break;
+		case OUTSIDE:
+			if (isCursorInsideBrowser) {
+				cursorPosition = CursorPosition.INSIDE;
+				sendMouseEvent(cursorLocationInControlCoordinate, SWT.MouseEnter);
+			}
+			break;
+		}
+		handleBrowserHover();
+	});
+}
+
+private enum CursorPosition {
+	INSIDE, OUTSIDE
+}
```

Comment:
I wonder whether it is worth to have an enum with only these two values, i.e., whether representing this via enums is more expressive/comprehensible than storing it in some `isCursorInsideBrowser` boolean.

In addition, I am not sure whether it is really necessary to store whether the cursor was inside the browser or whether we can just calculate it every 100ms, as it's only a cheap position-within-bounds check. That would reduce the number of instance fields that have to be managed thoroughly to not represent an inconsistent state.

Reply:
> In addition, I am not sure whether it is really necessary to store whether the cursor was inside the browser or whether we can just calculate it every 100ms...

Do you mean getting rid of the field `cursorPosition`?
If I read the code correctly, it is necessary to detect a `MouseEnter` / `MouseExit` situation. Calculating it on-the-fly would tell you were it is right now but you need to know where it was before in order to know if it entered/left an area.

Reply:
Yes, I mean getting rid of `cursorPosition`. There is `previousCursorLocation` from which you can always derive the value of `cursorPosition`, so it only serves as a cache. And since the calculation is so cheap, I would challenge the performance benefit of caching the value compared to the implicit dependency between the `previousCursorLocation` and `cursorPosition` values, necessity to update the together and the resulting risk of inconsistencies/regressions.

Reply:
Ah, I understand what you mean now.

I would rather leave it as it is right now. The reason for that is that if one uses `previousCursorLocation` to calculate the `cursorPosition` then the semantics become a bit (more) difficult and one also has to move the instruction `previousCursorLocation = cursorLocation;` below the `switch` statement because of the added "temporal dependency" (`previousCursorLocation` needs to remain **previous** before it becomes the **current (for the next iteration)** cursor location). 

Right now the only problem I see with the field is that its semantics change inside the `switch` statement: at the beginning it's the "last known position" (from the previous iteration) and inside the `switch`, where it is assigned, it becomes the "current position". But I prefer it that way, it makes the code easier to read.